catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswStore.java

```diff
@@ -0,0 +1,344 @@
+                .add(new InsertAction(insertTypeName, null, metacards));
+        try {
+            TransactionResponseType response = csw.transaction(transactionRequest);
+            Set<String> processedIds = new HashSet<>();
+            //dive down into the response to get the created ID's. We need these so we can query
+            //the source again to get the created metacards and put them in the result
+            createdMetacardFilters.addAll(response.getInsertResult()
+                    .stream()
+                    .map(InsertResultType::getBriefRecord)
+                    .flatMap(Collection::stream)
+                    .map(BriefRecordType::getIdentifier)
+                    .flatMap(Collection::stream)
+                    .map(JAXBElement::getValue)
+                    .map(SimpleLiteral::getContent)
+                    .flatMap(Collection::stream)
+                    .map(id -> {
+                        processedIds.add(id);
+                        return filterBuilder.attribute(Metacard.ID)
+                                .is()
+                                .equalTo()
+                                .text(id);
+                    })
+                    .collect(Collectors.toList()));
+            metacardIds.removeAll(processedIds);
+
+            for (String id : metacardIds) {
+                errors.add(new ProcessingDetailsImpl(id, null, "Failed to create metacard"));
+            }
+
+        } catch (CswException e) {
+            throw new IngestException("Csw Transaction Failed : ", e);
+        }
+
+        try {
+            createdMetacards.addAll(transactionQuery(createdMetacardFilters));
+        } catch (UnsupportedQueryException e) {
+            errors.add(new ProcessingDetailsImpl(this.getId(),
+                    e,
+                    "Failed to retrieve newly created metacards"));
+        }
+
+        return new CreateResponseImpl(createRequest, properties, createdMetacards, errors);
+    }
+
+    @Override
+    public UpdateResponse update(UpdateRequest updateRequest) throws IngestException {
+        Map<String, Serializable> properties = new HashMap<>();
+
+        validateOperation();
+
+        Subject subject =
+                (Subject) updateRequest.getPropertyValue(SecurityConstants.SECURITY_SUBJECT);
+        Csw csw = factory.getClientForSubject(subject);
+        CswTransactionRequest transactionRequest = getTransactionRequest();
+
+        OperationTransaction opTrans = (OperationTransaction) updateRequest.getPropertyValue(
+                Constants.OPERATION_TRANSACTION_KEY);
+
+        String insertTypeName = schemaTransformerManager.getTransformerIdForSchema(
+                cswSourceConfiguration.getOutputSchema());
+
+        HashSet<ProcessingDetails> errors = new HashSet<>();
+
+        if (insertTypeName == null) {
+            insertTypeName = CswConstants.CSW_RECORD;
+        }
+
+        ArrayList<Metacard> updatedMetacards = new ArrayList<>();
```

Comment:
If you move this list creation down to below the first for loop, you can initialize it with the correct size to avoid a resize.